firedrake/mesh.py

```diff
@@ -1010,6 +1114,152 @@ def _order_data_by_cell_index(self, column_list, cell_data):
         return cell_data[cell_list]
 
 
+# TODO: Could this be merged with MeshTopology given that dmcommon.pyx
+# now covers DMSwarms and DMPlexes?
+class VertexOnlyMeshTopology(AbstractMeshTopology):
+    """
+    Representation of a vertex-only mesh topology immersed within
+    another mesh.
+    """
+
+    def __init__(self, swarm, parentmesh, name, reorder):
+        """
+        Half-initialise a mesh topology.
+
+        :arg swarm: Particle In Cell (PIC) :class:`DMSwarm` representing
+            vertices immersed within a :class:`DMPlex` stored in the
+            `parentmesh`
+        :arg parentmesh: the mesh within which the vertex-only mesh
+            topology is immersed.
+        :arg name: name of the mesh
+        :arg reorder: whether to reorder the mesh (bool)
+        """
+
+        super().__init__(name)
+
+        # TODO: As a performance optimisation, we should renumber the
+        # swarm to in parent-cell order so that we traverse efficiently.
+        if reorder:
+            raise NotImplementedError("Mesh reordering not implemented for vertex only meshes yet.")
+
+        dmcommon.validate_mesh(swarm)
+        swarm.setFromOptions()
+
+        self._parent_mesh = parentmesh
+        self._topology_dm = swarm
+        self._comm = dup_comm(swarm.comm.tompi4py())
+
+        # A cache of shared function space data on this mesh
+        self._shared_data_cache = defaultdict(dict)
+
+        # Cell subsets for integration over subregions
+        self._subsets = {}
+
+        tdim = 0
+
+        cell = ufl.Cell("vertex")
+        self._ufl_mesh = ufl.Mesh(ufl.VectorElement("DG", cell, 0, dim=cell.topological_dimension()))
+
+        def callback(self):
+            """Finish initialisation."""
+            del self._callback
+
+            # Mark OP2 entities and derive the resulting Swarm numbering
+            with timed_region("Mesh: numbering"):
+                dmcommon.mark_entity_classes(self._topology_dm)
+                self._entity_classes = dmcommon.get_entity_classes(self._topology_dm).astype(int)
+
+                # Derive a cell numbering from the Swarm numbering
+                entity_dofs = np.zeros(tdim+1, dtype=IntType)
+                entity_dofs[-1] = 1
+
+                self._cell_numbering = self.create_section(entity_dofs)
+                entity_dofs[:] = 0
+                entity_dofs[0] = 1
+                self._vertex_numbering = self.create_section(entity_dofs)
+
+        self._callback = callback
```

Comment:
I think this callback dance is actually unnecessary. The reason for the callback stuff in `Mesh` is to do with not wanting to grow the overlap before multigrid refinement, but here we have none of this. So I think that you can just inline all of this stuff into the initialisation routine.

Reply:
Inlined the lot.

Reply:
It doesn't look inlined because you still have a callback. I meant to just do all the initialisation straight away. But please just do it next time round.